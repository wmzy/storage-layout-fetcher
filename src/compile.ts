import { spawn } from 'node:child_process';
import { Readable } from 'node:stream';

export default function compile(solcPath: string, input: string) {
  return new Promise<string>((resolve, reject) => {
    const solc = spawn(solcPath, ['--standard-json']);
      const streamToText = (s: Readable) =>
        s
          .setEncoding('utf8')
          .toArray()
          .then((chunks) => chunks.join(''));

         solc.stdin.end(input);
         const outPromise = streamToText(solc.stdout)
         const errorPromise = streamToText(solc.stderr)

    solc.once('close', (code) => {
      return code === 0 ? outPromise.then(resolve) : errorPromise.then(e => reject(new Error(e)));
    });
  });
}
