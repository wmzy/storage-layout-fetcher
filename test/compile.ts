import { readFile } from 'node:fs/promises';
import compile from '../src/compile';

describe('transform', () => {
  it('should compile success', async ({ expect }) => {
    const input = await readFile('test/fixtures/compile-input.json', 'utf-8');
    const result = compile('solc', input);
    await expect(result).resolves.toMatchSnapshot();
  });
});
