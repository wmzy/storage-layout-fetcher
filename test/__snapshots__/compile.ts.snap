// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transform > should compile success 1`] = `
"{"contracts":{"src/BaseRewardRateManager.sol":{"BaseRewardRateManager":{"storageLayout":{"storage":[{"astId":7174,"contract":"src/BaseRewardRateManager.sol:BaseRewardRateManager","label":"rateHistory","offset":0,"slot":"0","type":"t_array(t_struct(RateCheckpoint)7170_storage)dyn_storage"}],"types":{"t_array(t_struct(RateCheckpoint)7170_storage)dyn_storage":{"base":"t_struct(RateCheckpoint)7170_storage","encoding":"dynamic_array","label":"struct BaseRewardRateManager.RateCheckpoint[]","numberOfBytes":"32"},"t_struct(RateCheckpoint)7170_storage":{"encoding":"inplace","label":"struct BaseRewardRateManager.RateCheckpoint","members":[{"astId":7167,"contract":"src/BaseRewardRateManager.sol:BaseRewardRateManager","label":"timestamp","offset":0,"slot":"0","type":"t_uint256"},{"astId":7169,"contract":"src/BaseRewardRateManager.sol:BaseRewardRateManager","label":"rate","offset":0,"slot":"1","type":"t_uint256"}],"numberOfBytes":"64"},"t_uint256":{"encoding":"inplace","label":"uint256","numberOfBytes":"32"}}}}}},"sources":{"lib/openzeppelin-contracts-upgradeable/contracts/access/AccessControlUpgradeable.sol":{"id":0},"lib/openzeppelin-contracts-upgradeable/contracts/access/extensions/AccessControlEnumerableUpgradeable.sol":{"id":1},"lib/openzeppelin-contracts-upgradeable/contracts/proxy/utils/Initializable.sol":{"id":2},"lib/openzeppelin-contracts-upgradeable/contracts/utils/ContextUpgradeable.sol":{"id":3},"lib/openzeppelin-contracts-upgradeable/contracts/utils/PausableUpgradeable.sol":{"id":4},"lib/openzeppelin-contracts-upgradeable/contracts/utils/ReentrancyGuardUpgradeable.sol":{"id":5},"lib/openzeppelin-contracts-upgradeable/contracts/utils/introspection/ERC165Upgradeable.sol":{"id":6},"lib/openzeppelin-contracts/contracts/access/IAccessControl.sol":{"id":7},"lib/openzeppelin-contracts/contracts/access/extensions/IAccessControlEnumerable.sol":{"id":8},"lib/openzeppelin-contracts/contracts/interfaces/IERC1363.sol":{"id":9},"lib/openzeppelin-contracts/contracts/interfaces/IERC165.sol":{"id":10},"lib/openzeppelin-contracts/contracts/interfaces/IERC20.sol":{"id":11},"lib/openzeppelin-contracts/contracts/interfaces/draft-IERC6093.sol":{"id":12},"lib/openzeppelin-contracts/contracts/token/ERC20/ERC20.sol":{"id":13},"lib/openzeppelin-contracts/contracts/token/ERC20/IERC20.sol":{"id":14},"lib/openzeppelin-contracts/contracts/token/ERC20/extensions/IERC20Metadata.sol":{"id":15},"lib/openzeppelin-contracts/contracts/token/ERC20/utils/SafeERC20.sol":{"id":16},"lib/openzeppelin-contracts/contracts/utils/Address.sol":{"id":17},"lib/openzeppelin-contracts/contracts/utils/Context.sol":{"id":18},"lib/openzeppelin-contracts/contracts/utils/Errors.sol":{"id":19},"lib/openzeppelin-contracts/contracts/utils/Panic.sol":{"id":20},"lib/openzeppelin-contracts/contracts/utils/Pausable.sol":{"id":21},"lib/openzeppelin-contracts/contracts/utils/ReentrancyGuard.sol":{"id":22},"lib/openzeppelin-contracts/contracts/utils/introspection/IERC165.sol":{"id":23},"lib/openzeppelin-contracts/contracts/utils/math/Math.sol":{"id":24},"lib/openzeppelin-contracts/contracts/utils/math/SafeCast.sol":{"id":25},"lib/openzeppelin-contracts/contracts/utils/structs/EnumerableSet.sol":{"id":26},"src/BaseRewardRateManager.sol":{"id":27},"src/BaseStakingRewards.sol":{"id":28},"src/WithdrawRequestsManager.sol":{"id":29},"src/interfaces/IStakingRewards.sol":{"id":30}}}
"
`;
